source/assets/js/playground/autocomplete.ts

```diff
@@ -0,0 +1,241 @@
+  string: 'makes it easy to combine, search, or split apart strings',
+};
+
+const builtinModules: CompletionModule[] = moduleMembers.map(modMember => {
+  let variables;
+  if ('variables' in modMember) {
+    variables = modMember.variables;
+  }
+  return {
+    name: modMember.name,
+    description: moduleDescriptions[modMember.name],
+    functions: (modMember.functions ?? []).map(name => ({name})),
+    variables: (variables ?? []).map(name => ({name: `$${name}`})),
+  };
+});
+
+const moduleNameRegExp = new RegExp(`(${moduleNames.join('|')}).\\$?\\w*`);
+
+const moduleCompletions = Object.freeze(
+  builtinModules.map(mod => ({
+    label: `sass:${mod.name}`,
+    apply: `sass:${mod.name}`,
+    info: mod.description,
+    type: 'class',
+    validFor: identifier,
+  }))
+);
+
+// Completions for the import of built in modules, for instance "sass:color".
+function builtinModulesCompletion(
+  context: CompletionContext
+): CompletionResult | null {
+  const nodeBefore = syntaxTree(context.state).resolveInner(context.pos, -1);
+  const potentialTypes = [
+    'StringLiteral', // When wrapped in quotes- `"sass:"`
+    'ValueName', // No end quote, before semicolon- `"sa`
+    ':', // No end quote, on semicolon- `"sass:`
+    'PseudoClassName', // No end quote, after semicolon- `"sass:m`
+  ];
+  if (!potentialTypes.includes(nodeBefore.type.name)) return null;
+  const potentialParentTypes = [
+    'UseStatement', // When wrapped in quotes
+    'PseudoClassSelector', // No end quote, after the colon
+  ];
+  if (!potentialParentTypes.includes(nodeBefore.parent?.type.name || ''))
+    return null;
+
+  const moduleMatch = context.matchBefore(/["'](sass:)?\w*/);
+
+  if (!moduleMatch) return null;
+  if (moduleMatch.from === moduleMatch.to && !context.explicit) return null;
+  return {
+    from: moduleMatch.from + 1,
+    to: moduleMatch.to,
+    options: moduleCompletions,
+    validFor: identifier,
+  };
+}
+
+const moduleVariableCompletions = Object.freeze(
+  builtinModules.reduce(
+    (acc: {[k: string]: CompletionResult['options'] | []}, mod) => {
+      acc[mod.name] =
+        mod.variables?.map(variable => ({
+          label: `${mod.name}.${variable.name}`,
+          info: variable?.description,
+          type: 'variable',
+        })) || [];
+      return acc;
+    },
+    {}
+  )
+);
+
+const moduleFunctionsCompletions = Object.freeze(
+  builtinModules.reduce(
+    (acc: {[k: string]: CompletionResult['options'] | []}, mod) => {
+      acc[mod.name] =
+        mod.functions?.map(variable => ({
+          label: `${mod.name}.${variable.name}`,
+          apply: `${mod.name}.${variable.name}(`,
+          info: variable?.description,
+          type: 'method',
+          boost: 10,
+          validFor: identifier,
+        })) || [];
+      return acc;
+    },
+    {}
+  )
+);
+
+// Returns the list of built in modules that are included in the text.
+function includedBuiltinModules(state: EditorState) {
```

Comment:
Since the editor does seem to maintain some degree of a syntax tree, is it possible to do this in a slightly more robust way? It's not a big deal to have this have false positives, but if it's easy to iterate through top-level AST nodes looking for `@use` that might be better.

Reply:
I switched to this in 55932c5feebbba79b253de2fa37cdfb5966d77a2.